### PR TITLE
Add Nitrokey FIDO U2F entry

### PR DESCRIPTION
**Pull request by alex-nitrokey**
This entry for the Nitrokey FIDO U2F is the same like in [libu2f-host](https://github.com/Yubico/libu2f-host/blob/master/70-u2f.rules) and is working just fine. It should be enough to install libnitrokey to get the device working in the future.

See [this discussion](https://support.nitrokey.com/t/problems-with-nitrokey-fido-u2f/1553) as well.

### CODE REVIEW

File: data/41-nitrokey.rules [hunk 1]
```diff
@@ -1,5 +1,7 @@
 # Nitrokey U2F
```
Comment (szszszsz) [inline data/41-nitrokey.rules]: Please add guard for the U2F devices, like in:
```
ACTION!="add|change", GOTO="u2f_end"
<here U2F devices>
LABEL="u2f_end"
```

File: data/41-nitrokey.rules [hunk 2]
```diff
@@ -1,5 +1,7 @@
...
 KERNEL=="hidraw*", SUBSYSTEM=="hidraw", MODE="0664", GROUP="plugdev", ATTRS{idVendor}=="2581", ATTRS{idProduct}=="f1d0"
```
Comment (szszszsz) [inline data/41-nitrokey.rules]: Please rewrite old U2F device rule to the new style

File: data/41-nitrokey.rules [hunk 3]
```diff
@@ -1,5 +1,7 @@
...
+# Nitrokey FIDO U2F
+KERNEL=="hidraw*", SUBSYSTEM=="hidraw", ATTRS{idVendor}=="20a0", ATTRS{idProduct}=="4287", TAG+="uaccess"
```
Comment (szszszsz) [inline data/41-nitrokey.rules]: Please check, should we not use `SUBSYSTEM=="usb"`, instead of `hidraw`.

Comment (alex-nitrokey) [reply data/41-nitrokey.rules]: Why do you think we should? We did not use usb subsystem in any other rule. Furthermore, no [other U2F device](https://github.com/Yubico/libu2f-host/blob/master/70-u2f.rules) seems to use it.

Comment (szszszsz) [reply data/41-nitrokey.rules]: I am sorry - I do not remember anymore the reason. I guess I thought about greater generalization for the rule. Please skip this for now.

File: data/41-nitrokey.rules [hunk 4]
```diff
@@ -1,5 +1,35 @@
...
+# SPDX-License-Identifier: LGPL-3.0
+#
+
+
+# this udev file should be used with udev 188 and newer
+ACTION!="add|change", GOTO="u2f_end"
+
 # Nitrokey U2F
-KERNEL=="hidraw*", SUBSYSTEM=="hidraw", MODE="0664", GROUP="plugdev", ATTRS{idVendor}=="2581", ATTRS{idProduct}=="f1d0"
+KERNEL=="hidraw*", SUBSYSTEM=="hidraw", ATTRS{idVendor}=="2581", ATTRS{idProduct}=="f1d0", MODE="0664", GROUP+="plugdev", TAG+="uaccess"
```
[the remark this answers is not in the document]
Comment (szszszsz) [reply data/41-nitrokey.rules]: This seems to be a complete documentation: http://reactivated.net/writing_udev_rules.html
Mode is for setting `chmod` over a device file, after a connection. `664` means r/w access for user and group.

**Comment by alex-nitrokey** [comment]
I tested f88c44f with Ubuntu 16.04 and Arch Linux. Ubuntu 18.04 should work fine as users already said it works.

What do you think, is it enough to include it? What else system would you like to get tested?

**Comment by alex-nitrokey** [comment]
btw: I found it difficult to find up-to-date overview information about udev. Did you find any good ressource? The one you linked above is very old and thus does not include uaccess at all.

**Comment by szszszsz** [comment]
Surprisingly to me, I cannot find any fresh materials.
The most expanded seem to be the one on Arch Wiki: https://wiki.archlinux.org/index.php/udev.

---------------------------
I have just found interesting issues, which discourages using static groups, encouraging `uaccess` only. Please take a look into them, and their references as well: 
- https://github.com/systemd/systemd/issues/4288
- https://github.com/OpenCBM/OpenCBM/pull/8

Some of the articles mentioned in the issues:
- https://enotty.pipebreaker.pl/2012/05/23/linux-automatic-user-acl-management/
- https://lintian.debian.org/tags/udev-rule-missing-uaccess.html

**Comment by alex-nitrokey** [comment]
Thanks a lot for this extensive collection. I hope some persons will find this in future. So my impression that it is difficult to find information about it wasn't wrong :smile:

Still, I did not see any hint that it would be harmful to have the group tag *additionally* to the uaccess rule. Nevertheless, it seems like the uaccess is supported for a very long time so that we can rely on it. Do you think I should remove the GROUP and MODE tag?

**Comment by alex-nitrokey** [comment]
Just a side note: We are using GROUP and uaccess for a long time for all other models without any problems so far. If most users are not in group plugdev than nothing happens anyway.

**Comment by szszszsz** [comment]
You are welcome! Right, it was not easy to find that. I think someone wrote in the linked urls, that using `plugdev` group is a security vulnerability. If that is so, and if the `uaccess` is in wide use, I would left only this one.

What was the main difference again between our udev rules files, which helped some of the users on the forum? I believe that the `uaccess` addition has helped, if I am not wrong.

Please confirm both.

**Comment by alex-nitrokey** [comment]
Yes, the uaccess helped, without it doesn't worked as the plugdev group does not exist on many systems nowadays.

The way I am seeing it, having the option in is only malicous (in any way) if the plugdev group does exist and the user is member of this group. (if at all) This is an "if" which should not be the case for systems that needs the uaccess tag anyway.

Still I would prefer removing the group and mode part if we know that all major distribution do still work. We might do something similar like the libu2f-host people by having an "old" file for the few people still needing it but release the other one in general. It seems to work there.

**Comment by szszszsz** [comment]
Agreed. I, on the other hand, see this as increasing the potential attack surface. 

Please prepare the 2 Udev rules' files then, with old and new way of writing. We will distribute the latter by default.

**Comment by szszszsz** [comment]
Please let me know, when this would ready for review. As of now it looks good in both -old and 'new' cases.

**Comment by alex-nitrokey** [comment]
Yes, this should be fine indeed.

**Comment by alex-nitrokey** [comment]
Please merge if you think it is fine so that I can reflect that on the website.

**Comment by szszszsz** [comment]
Sorry for delay. I have corrected the rules to be more consistent. Could you review my changes please, and test rules as a whole?

**Comment by alex-nitrokey** [comment]
I can not test the old rules (this is why I left them as is but I am quite sure that they will work). I don't know which system does still use this.

The new rules file did not change by your commits. So this should be fine. Your changes look fine for me. I just hesitated to change the files we already used successfully, but as the changes are based on (minor) security considerations it makes sense for me to change it the way you did. Thanks.

Can not approve it via the Github panel.

**Comment by szszszsz** [comment]
Great! Merging.

Another note regarding `plugdev`: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=897916

**Comment by alex-nitrokey** [comment]
thanks, will change references online

**Pull request closed without merging**